### PR TITLE
GRPO Loss kernel fully write by triton, reduce 46G memory 

### PR DESCRIPTION
**Pull request by mdy666**
## Summary

This is grpo_loss kernel for GRPOTrainer in TRL, it fully writes by triton. This kernel offers higher precision(`close fp32`), faster speed(`10x`), and lower GPU memory consumption(`reduce 46G`). 
This is my repo: https://github.com/mdy666/mdy_triton/tree/main/triton_kernels/grpo
And i provide a code that run a micro batch one by one, maybe it can reduce memory to 1 / mbs of the original. It dose not need chunk loss

## fused_selective_log_softmax
It compue old_logp and ref_logp, it reduce 10G peak Memory. it does not requires grad and output fp32 logp for higher precision.

**Precision**
```bash
test/transformers/test_grpo_loss.py::test_selective_log_softmax[dtype0-1e-05-1e-05-0.9-1-1024-64000] 
==================== selective_log_softmax ====================
torch-bf16 vs torch-fp32, Max difference: 0.003912173677235842, Mean difference: 0.0013292559888213873
triton-bf16 vs torch-fp32, Max difference: 1.887589462512551e-07, Mean difference: 1.112928593727247e-08
PASSED
test/transformers/test_grpo_loss.py::test_selective_log_softmax[dtype0-1e-05-1e-05-0.7-1-1024-151936] 
==================== selective_log_softmax ====================
torch-bf16 vs torch-fp32, Max difference: 0.004012447316199541, Mean difference: 0.0012363159330561757
triton-bf16 vs torch-fp32, Max difference: 1.3388500974542694e-07, Mean difference: 1.2898424550655818e-08
PASSED
```

**Speed**
only need forward (ms)
image
```bash
   vocab_size    triton      torch
0     32000.0  0.514511   4.622731
1     64000.0  1.024138   8.693873
2     96000.0  1.477454  13.541513
3    128000.0  2.022543  18.250278
4    160000.0  2.582751  22.690176
5    192000.0  2.950109  27.426208
```

**Memory**
It can ruduce lots of peak memory. Because it does not save `logits / temperature` in HBM. For `bs=16 output_length=2048 vocab_size=151936`. It can reduce peak memoy from 20G to 10G

## triton_grpo_loss
It compue the grpo loss fully by triton kernel, it does not creates any redundant memory.

**Precision**
```bash
test/transformers/test_grpo_loss.py::test_grpo_loss[dtype0-1e-05-1e-05-1-1024-151936-0.7-1-0.0-0.2-0.4] 
==================== grpo_loss ====================
`num_iteration=1 beta=0.0 eps_low=0.2 eps_high=0.4`
per_token_loss: torch-bf16 vs torch-fp32, Max difference: 0.0, Mean difference: 0.0
logits.grad: torch-bf16 vs torch-fp32, Max difference: 0.04707478731870651, Mean difference: 0.005103979725390673
per_token_loss: triton-bf16 vs torch-fp32, Max difference: 0.0, Mean difference: 0.0
logits.grad: triton-bf16 vs torch-fp32, Max difference: 0.003846844658255577, Mean difference: 0.0004555800405796617
PASSED
test/transformers/test_grpo_loss.py::test_grpo_loss[dtype0-1e-05-1e-05-1-1024-151936-0.7-1-0.04-0.2-0.4] 
==================== grpo_loss ====================
`num_iteration=1 beta=0.04 eps_low=0.2 eps_high=0.4`
per_token_loss: torch-bf16 vs torch-fp32, Max difference: 0.06327077001333237, Mean difference: 0.014040875248610973
per_token_kl: torch-bf16 vs torch-fp32, Max difference: 0.06327081471681595, Mean difference: 0.01404162310063839
logits.grad: torch-bf16 vs torch-fp32, Max difference: 0.11991540342569351, Mean difference: 0.018816275522112846
per_token_loss: triton-bf16 vs torch-fp32, Max difference: 2.855108732546796e-06, Mean difference: 5.513994665307109e-07
per_token_kl: triton-bf16 vs torch-fp32, Max difference: 2.7946493901254144e-06, Mean difference: 5.514012855201145e-07
logits.grad: triton-bf16 vs torch-fp32, Max difference: 0.003891229862347245, Mean difference: 0.0012611059937626123
PASSED
test/transformers/test_grpo_loss.py::test_grpo_loss[dtype0-1e-05-1e-05-1-1024-151936-0.7-5-0.0-0.2-0.4] 
==================== grpo_loss ====================
`num_iteration=5 beta=0.0 eps_low=0.2 eps_high=0.4`
per_token_loss: torch-bf16 vs torch-fp32, Max difference: 0.0, Mean difference: 0.0
logits.grad: torch-bf16 vs torch-fp32, Max difference: 0.0, Mean difference: 0.0
per_token_loss: triton-bf16 vs torch-fp32, Max difference: 0.0, Mean difference: 0.0
logits.grad: triton-bf16 vs torch-fp32, Max difference: 0.0, Mean difference: 0.0
PASSED
test/transformers/test_grpo_loss.py::test_grpo_loss[dtype0-1e-05-1e-05-1-1024-151936-0.7-5-0.04-0.2-0.4] 
==================== grpo_loss ====================
`num_iteration=5 beta=0.04 eps_low=0.2 eps_high=0.4`
per_token_loss: torch-bf16 vs torch-fp32, Max difference: 0.058179255574941635, Mean difference: 0.01415303349494934
per_token_kl: torch-bf16 vs torch-fp32, Max difference: 0.05817938596010208, Mean difference: 0.01415514200925827
logits.grad: torch-bf16 vs torch-fp32, Max difference: 0.11716917902231216, Mean difference: 0.018849259242415428
per_token_loss: triton-bf16 vs torch-fp32, Max difference: 2.6981190330843674e-06, Mean difference: 5.632235797747853e-07
per_token_kl: triton-bf16 vs torch-fp32, Max difference: 2.68613439402543e-06, Mean difference: 5.63439357392781e-07
logits.grad: triton-bf16 vs torch-fp32, Max difference: 0.0038916049525141716, Mean difference: 0.0012580838520079851
PASSED
```

**Speed**
forward + backward (ms)
image

```bash
   vocab_size    triton      torch
0     32000.0  1.496177  17.489887
1     64000.0  2.967113  33.858929
2     96000.0  4.331829  47.953056
3    128000.0  5.840693  63.874176
4    160000.0  7.363397  79.620605
5    192000.0  8.956139  95.290466
```

**Memory**
It can ruduce lots of memory. Because it does not save `logits / temperature` and `log_softmax_logits` in HBM and use `logits.data` to store 'logits.grad'. For `bs=16 output_length=2048 vocab_size=151936`. It can reduce memoy from 56G  to 10G.
The following picture illustrates the principle of the algorithm.
89b792484c6f8e66b1116e7ab8152ff

## Environment
torch2.6-cuda12.6
triton 3.2
GPU：H200

### CODE REVIEW

**Comment by austin362667** [comment]
Nice work! Thank you @mdy666 
Is this triton-based GRPO benchmarked with torch-based GRPO (in TRL) or torch-compiled GRPO (in [Liger](https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/chunked_loss/grpo_loss.py))

**Comment by mdy666** [comment]
> Nice work! Thank you @mdy666 Is this triton-based GRPO benchmarked with torch-based GRPO (in TRL) or torch-compiled GRPO (in [Liger](https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/chunked_loss/grpo_loss.py))

It benchmarked with torch-based GRPO (in TRL). It optimize the `_get_per_token_logps` and `compute_loss ` function in GRPOTrainer and easy to use.

**Comment by mdy666** [comment]
I test the peak memory use the flowing code compared with LIgerGRPO
## My
```python
vocab_size = 151936
B, L = 4, 2048
device = "cuda"
dtype = torch.bfloat16
logits = torch.randn(B, L + 1, vocab_size, device=device, dtype=dtype)
logits.requires_grad_(True)

completion_ids = torch.randint(0, vocab_size-1, (B, L), dtype=torch.int64, device=device)
completion_mask = None
ref_logp = torch.randn(B, L, device=device, dtype=torch.float32)
# ref_logp = None
old_logp = torch.randn(B, L, device=device, dtype=torch.float32)
# old_logp = None
dy = torch.randn_like(ref_logp)
advantages = torch.randn(B, device=device, dtype=torch.float32)
temperature, beta, eps_low, eps_high = 0.9, 0.1, 0.5, 1.5
time.sleep(5)

for _ in tqdm(range(50)):
    loss = triton_grpo_loss(logits,
                old_logp,
                ref_logp,
                completion_ids,
                advantages,
                completion_mask,
                temperature,
                beta,
                eps_low,
                eps_high)[0]
    loss.backward(dy)
    logits.grad = None
```

## Liger
```python
from liger_kernel.chunked_loss import LigerFusedLinearGRPOLoss
loss_fn = LigerFusedLinearGRPOLoss(compiled=True, 
                                   chunk_size=1,
                                   temperature=0.9)

vocab_size = 151936
hidden_size = 1536
B, L = 4, 2048
device = "cuda"
dtype = torch.bfloat16
_input = torch.randn(B, L, hidden_size, device=device, dtype=dtype)
fc = torch.nn.Linear(hidden_size, vocab_size, device=device, dtype=dtype)
lin_weight = fc.weight

completion_ids = torch.randint(0, vocab_size-1, (B, L), dtype=torch.int64, device=device)
completion_mask = torch.ones_like(completion_ids)
ref_logp = torch.randn(B, L, device=device, dtype=torch.float32)
# ref_logp = None
old_logp = torch.randn(B, L, device=device, dtype=torch.float32)
# old_logp = None
dy = torch.randn_like(ref_logp)
advantages = torch.randn(B, device=device, dtype=torch.float32)

for _ in tqdm(range(50)):
    loss, _ = loss_fn(_input,
                   lin_weight,
                   completion_ids,
                   completion_mask,
                   advantages,
                 ref_per_token_logps=ref_logp,
                old_per_token_logps=old_logp,
                   )
    loss.backward()
```

## Compare
if set `batch_size=16`, my code memory is from 10G to 10G, liger is 1G -> 4.6G
if set `batch_size=4`, my code memory is from 2.9G to 2.9G, liger is 1G -> 4.6G

For example if I split a micro bacth(mbs=16) to 4 chunk, my kernel save more memory and the activation become 1/4 to original, and the GPU utilization only decrease a little. And maybe don't need set `gradient_checkpoint=True`, it will be more fast

**Comment by shivam15s** [comment]
Hey @mdy666 , great work! Curious about the loss mismatch: we also do reduction using `completion_mask.sum()` here: https://github.com/linkedin/Liger-Kernel/blob/908b89c4dc9bb872351887b382a1e09ca25fbe85/src/liger_kernel/chunked_loss/grpo_loss.py#L75

**Comment by mdy666** [comment]
> Hey @mdy666 , great work! Curious about the loss mismatch: we also do reduction using `completion_mask.sum()` here:
> 
> https://github.com/linkedin/Liger-Kernel/blob/908b89c4dc9bb872351887b382a1e09ca25fbe85/src/liger_kernel/chunked_loss/grpo_loss.py#L75

My description may not be accurate. I upload the `test code` again. In `TRL-GRPO` it always set `per_device_batch_size(mbs)=num_generation`, like 8 or 16,  it is very big compared with `sft` for LLM. So it' s a good way to split the `mbs` to some chunks for model forward and backward when the `advantages` is computed. In `LigerFusedLinearGRPOLoss`, it output a `sclar loss` which is reduced by `completion_mask[idx:idx+STEP].sum()`, it should use `completion_mask.sum()` to reduce, like `num_items_in_batch` in `cross_entropy_loss`. If split `per_device_batch_size(mbs)` to some chunks, it should do a samll change for `LigerFusedLinearGRPOLoss`

**Comment by shivam15s** [comment]
Thanks for your comment, we reduce using `full_attention_mask.sum()` (ie `completion_mask.sum()` instead of `completion_mask[start:end].sum()`)

**Comment by mdy666** [comment]
> Thanks for your comment, we normalize using `full_attention_mask.sum()` (ie `completion_mask.sum()` instead of `completion_mask[start:end].sum()`)



> Thanks for your comment, we normalize using `full_attention_mask.sum()` (ie `completion_mask.sum()` instead of `completion_mask[start:end].sum()`)

No. I spilit the micro_batch. In `LigerFusedLinearGRPOFunction`,  the `full_attention_mask` is `completion_mask[start:end]` and `attention_mask` is `completion_mask[start:end][index]`
```python
elif loss_type == "bnpo":
    # Batch Normalized Per-token loss (original implementation)
    loss = (per_token_loss * attention_mask).sum() / torch.clamp(full_attention_mask.sum(), min=1.0)
```

**Comment by shivam15s** [comment]
Oh gotcha, your approach has a custom splitting into micro batches which would require a different way of normalizing 👍

**Review by lancerts — APPROVED** [review]
